pybigquery/sqlalchemy_bigquery.py

```diff
@@ -360,6 +360,10 @@ def visit_notendswith_op_binary(self, binary, operator, **kw):
 
     __expanded_param = re.compile(fr"\(\[" fr"{__expandng_text}" fr"_[^\]]+\]\)$").match
 
+    __remove_type_parameter = _helpers.substitute_re_method(
+        r"(\w+)\(\s*\d+\s*(,\s*\d+\s*)*\)", r"\1"
+    )
+
```

Comment:
I tried the verbose syntax the other day at @plamut's suggestion and I couldn't get it right. :shrug: Given that I have a lot of regex experience, I think that's a bad sign.

The biggest issue for these regexes is the mixture of literal parentheses and square brackets.

Reply:
FWIW, I find reading the terse forms is a big speedbump.  The terse form is confusing, because it discards the precision bit, although it works for just ripping off the base type:

```python
>>> import re
>>> terse = r"(\w+)\(\s*\d+\s*(,\s*\d+\s*)*\)"
>>> re_terse = re.compile(terse)
>>> re_terse.match("NUMERIC(1)").groups()
('NUMERIC', None)
>>> re_terse.match("NUMERIC(1,3)").groups()
('NUMERIC', ',3')
```

The verbose form with the same behavior:

```python
>>> verbose = r"""
...     (\w+)          # base type
...     \(             # literal open paren
...     \s*            # whitespace
...     \d+            # length / precision
...     \s*            # whitespace
...     (              # optional scale group
...     ,              # literal comma
...     \s*            # whitespace
...     \d+            # scale
...     \s*            # whitespace
...     )*             # close scale group
...     \)             # literal close paren
... """
>>> re_verbose = re.compile(verbose, re.VERBOSE)
>>> re_verbose.match("NUMERIC(1)").groups()
('NUMERIC', None)
>>> re_verbose.match("NUMERIC(1 ,3)").groups()
('NUMERIC', ',3')
```

For future readability, I'd be inclined to have compiled versions for each of the main cases (`STRING`, `BYTES`, `{,BIG}NUMERIC`) which gave back named groups, although I realize that using them in this particular case would be trickier.

Reply:
I like what you did there.  I went all in :), although I didn't break things down that far.